pkg/controller/workload/bpfcache/locality_cache.go

```diff
@@ -0,0 +1,178 @@
+	nodeName  string // init from os.Getenv("NODE_NAME"), workload.GetNode()
+	clusterId string // init from workload.GetClusterId()
+	network   string // workload.GetNetwork()
+	mask      uint32 // mask
+}
+
+func Valid(s string) bool {
+	return s != ""
+}
+
+func (l *localityInfo) Set(s string, param uint32) {
+	if !Valid(s) {
+		return
+	}
+	switch param {
+	case REGION:
+		l.region = s
+	case ZONE:
+		l.zone = s
+	case SUBZONE:
+		l.subZone = s
+	case NODENAME:
+		l.nodeName = s
+	case CLUSTERID:
+		l.clusterId = s
+	case NETWORK:
+		l.network = s
+	}
+	l.mask |= param
+}
+
+func (l *localityInfo) Clear(param uint32) {
+	l.mask &= ^param
+}
+
+func (l *localityInfo) IsSet(param uint32) bool {
+	return l.mask&param != 0
+}
+
+type LocalityCache struct {
+	mutex                  sync.RWMutex
+	localityInfo           localityInfo
+	isLocalityInfoSet      bool
+	RoutingPreference      []workloadapi.LoadBalancing_Scope
+	routePreferenceCount   uint32
+	isRoutingPreferenceSet bool
+	workloadWaitQueue      map[*workloadapi.Workload]struct{}
+}
+
+func NewLocalityCache() *LocalityCache {
+	return &LocalityCache{
+		localityInfo:           localityInfo{},
+		isLocalityInfoSet:      false,
+		RoutingPreference:      make([]workloadapi.LoadBalancing_Scope, 0),
+		isRoutingPreferenceSet: false,
+		workloadWaitQueue:      make(map[*workloadapi.Workload]struct{}),
+	}
+}
+
+func (l *LocalityCache) SetLocality(nodeName, clusterId, network string, locality *workloadapi.Locality) {
+	// notice: nodeName should set by processor or os.Getenv("NODE_NAME"),
+	l.localityInfo.Set(nodeName, NODENAME)
+	l.localityInfo.Set(locality.GetRegion(), REGION)
+	l.localityInfo.Set(locality.GetSubzone(), SUBZONE)
+	l.localityInfo.Set(locality.GetZone(), ZONE)
+	l.localityInfo.Set(clusterId, CLUSTERID)
+	l.localityInfo.Set(network, NETWORK)
+
+	l.isLocalityInfoSet = true
+}
+
+func (l *LocalityCache) SetRoutingPreference(s []workloadapi.LoadBalancing_Scope) {
+	// notice: s should set by lb.GetRoutingPreference()
+	if len(s) > 0 {
+		l.RoutingPreference = s
+		l.isRoutingPreferenceSet = true
+		l.routePreferenceCount = uint32(len(s))
+	}
+}
+
+func (l *LocalityCache) CanLocalityLB() bool {
+	log.Debugf("isLocalityInfoSet: %#v, isRoutingPreferenceSet: %#v", l.isLocalityInfoSet, l.isRoutingPreferenceSet)
+	return l.isLocalityInfoSet && l.isRoutingPreferenceSet
+}
+
+func (l *LocalityCache) CalcuLocalityLBPrio(wl *workloadapi.Workload) uint32 {
+	var rank uint32 = 0
+	for scope := range l.RoutingPreference {
+		switch scope {
+		case int(workloadapi.LoadBalancing_REGION):
+			log.Debugf("l.localityInfo.IsSet(REGION) %#v, Valid(wl.GetLocality().GetRegion()) %#v, l.localityInfo.region %#v, wl.GetLocality().GetRegion() %#v", l.localityInfo.IsSet(REGION), Valid(wl.GetLocality().GetRegion()), l.localityInfo.region, wl.GetLocality().GetRegion())
+			if l.localityInfo.IsSet(REGION) && Valid(wl.GetLocality().GetRegion()) && l.localityInfo.region == wl.GetLocality().GetRegion() {
```

Comment:
```suggestion
			if l.localityInfo.region == wl.GetLocality().GetRegion() {
```

Reply:
This looks not right, think of a case only `LoadBalancing_REGION` is set, if the workload has same region with local node, the rank(priority) should be 0, otherwise 1

Reply:
We should only deal with fields that have been setup successfully. The default value of locality such as Region is "". If we don't check weather the value has been set, We will get rank=1 while workload comes with unset resion filed.  Also, If the Node is label with locality, the upcomming workload will have this fileds set, and `l.localityInfo.IsSet(REGION)` is always been set.